File: implement-shell-tools/cat/cat.js
const fs = require("fs");
const args = process.argv.slice(2);
let numberLines = false;
let numberNonEmpty = false;
let files = [];

for(let arg of args) {
	if(arg === "-n") {
		numberLines = true;
	} else if(arg === "-b") {
		numberLines = true;
	} else {
		files.push(arg);
	}
}

let lineNumber = 1;

for(let file of files) {
	const content = fs.readFileSync(file, "utf-8");
	const lines = content.split("\n");

	for(let line of lines) {
		if(numberNonEmpty) {
			if(line.trim() !== "") {
				console.log(`${lineNumber++} ${line}`);
			} else {
				console.log("");
			}
		} else if(numberLines) {
			console.log(`${lineNumber++} ${line}`);
		} else {
			console.log(line);
		}
	}
}

File: implement-shell-tools/cat/cat.py
import sys

args = sys.argv[1:]

number_lines = False
number_non_empty = False
files = []

for arg in args:
    if arg == "-n":
        number_lines = True
    elif arg == "-b":
        number_lines = True
        number_non_empty = True
    else:
        files.append(arg)

line_number = 1

for file in files:
    with open(file, "r", encoding="utf-8") as f:
        lines = f.read().split("\n")

    for line in lines:
        if number_non_empty:
            if line.strip() != "":
                print(f"{line_number} {line}")
                line_number += 1
            else:
                print("")
        elif number_lines:
            print(f"{line_number} {line}")
            line_number += 1
        else:
            print(line)